Proxy initial commit

bergwolf opened the pull request.

bergwolf commented on Makefile:
```diff
@@ -0,0 +1,10 @@
+all:
```
Do we need to have apache 2 licensing statement in every file? There is already an apche LICENSE file in the top directory and I don't think we will use other licenses in the project, right?

OTOH, if it is an Intel open source project policy, I'm willing to follow it.

grahamwhaley replied:
To add to that - I'll note that we are looking at using 'SPDX' style license headers for these new projects - a much smaller and cleaner method than placing the full or part license text in each file.
For reference:
https://github.com/kata-containers/ksm-throttler/pull/1/files#diff-360cbfd0e0c83e21065226e59696f652R5
https://spdx.org/using-spdx

bergwolf replied:
SPDX sounds good to me.

bergwolf commented on Makefile:
```diff
@@ -0,0 +1,10 @@
+	go build proxy.go
+	make -C test
+
+test: all
```
I would prefer `make` command w/o arguments to just build the project rather than running tests altogether. That is also how most Makefiles work.

jodh-intel replied:
Ah - yes, sorry. It hadn't registered that `all` is the default rule. I guess it's fine as-is or you could make it:

```
default: build

all: default test
```

bergwolf commented on proxy.go:
```diff
@@ -0,0 +1,83 @@
+	"net"
+	"sync"
+
+	"github.com/golang/glog"
```
All hyper projects use `glog` instead. We should make a decision on which log system to use for kata and migrate to it.

jodh-intel replied:
Agreed. What's your view @sameo, @grahamwhaley? Do we need to document this somewhere maybe?

bergwolf replied:
I've taken the question to the slack channel. Let's discuss it there.

bergwolf commented on proxy.go:
```diff
@@ -0,0 +1,83 @@
+	"github.com/hashicorp/yamux"
+)
+
+// @channel is the unix socket address we want to multiplex
```
Yep, it was left over from previous versions. I'll just drop it.

jodh-intel replied:
Great - thanks.

bergwolf commented on proxy.go:
```diff
@@ -0,0 +1,83 @@
+
+func main() {
+	channel := flag.String("s", "/tmp/target.sock", "unix socket to multiplex on")
+	proxyAddr := flag.String("l", "/tmp/proxy.sock", "unix socket to listen at")
```
Good point, I'll add URI support here. And I agree that default paths are useless.

bergwolf commented on test/test-cmd.sh:
```diff
@@ -0,0 +1,38 @@
+		output=$(./client -f ${f})
+		result=$(echo ${output}|grep SUCCESS 2>/dev/null || true)
+		if [ x"${result}" == "x" ]; then
+			echo test failed with ${output}
```
The error message from `grep` is useless. We just need to know if `grep` succeeds or not.

bergwolf commented on test/test-cmd.sh:
```diff
@@ -0,0 +1,38 @@
+# start server
+./server &
+# sleep a bit to let server spin up
+sleep 2
```
I tend to think that busy pulling with a timeout is a bit over-skill just for this little test script.

jodh-intel replied:
Yes, it's a very simple script at present, so I don't think this is a "blocker" to get this landed. It's just something we need to think about going forward as the script grows to avoid odd failures on (very ;) slow systems.

bergwolf replied:
Overall the `proxy` program will be very simple & small. But yeah, who knows ;)

grahamwhaley replied:
Yeah, we have seen problems with fixed timeouts (even quite large timeouts) when running CI tests on cloud based VM machines - so we've become wary of fixed timeouts and try to avoid them when we can. Flaky CI failures can become the bane of our lives otherwise :-)

bergwolf commented on test/test-cmd.sh:
```diff
@@ -0,0 +1,38 @@
+pkill proxy
+
+rm -f /tmp/proxy.sock /tmp/target.sock
+echo test result is $?
```
I was using it as a indicator that the test is done because otherwise there is no output. But you do have a valid point that `$?` is redundant here.

bergwolf commented:
updated and CI green. @jodh-intel @grahamwhaley PTAL.

sameo commented:
@bergwolf I assume you're going to rebase this PR based on the final gRPC definition? Or do you want us to give it a final review and merge it without gRPC support first?

bergwolf commented:
@sameo please review it as is. The proxy does not rely on gRPC support.

sameo commented:
@bergwolf ah right, your implementation copies all frames coming from the shim into the yamux endpoint, so it's not a gRPC server itself. I guess doing otherwise would be more secure but also much more complex.

sameo commented on proxy.go:
```diff
@@ -0,0 +1,100 @@
+			glog.Errorf("fail to accept new connection: %s", err)
+			return err
+		}
+		stream, err := session.Open()
```
Could you please put the whole stream opening, stream copy and sync logic into a dedicated function and call a go routine on it for each accept? I think this would make the code more readable.

bergwolf commented:
updated to address @sameo's comments and added unit tests. glog is also replaced with logrus. We'll add a wrapper above logrus to print glog format logs when need.

sameo commented:
@bergwolf Thanks for the update.
LGTM
